src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/JNativeScanTask.java

```diff
@@ -0,0 +1,193 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.tools.jnativescan;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.lang.constant.ClassDesc;
+import java.lang.module.Configuration;
+import java.lang.module.ModuleFinder;
+import java.lang.module.ResolvedModule;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipFile;
+
+class JNativeScanTask {
+
+    private final PrintWriter out;
+    private final List<Path> classPaths;
+    private final List<Path> modulePaths;
+    private final List<String> cmdRootModules;
+    private final Runtime.Version version;
+    private final Action action;
+
+    public JNativeScanTask(PrintWriter out, List<Path> classPaths, List<Path> modulePaths,
+                           List<String> cmdRootModules, Runtime.Version version, Action action) {
+        this.out = out;
+        this.classPaths = classPaths;
+        this.modulePaths = modulePaths;
+        this.version = version;
+        this.action = action;
+        this.cmdRootModules = cmdRootModules;
+    }
+
+    public void run() throws JNativeScanFatalError {
+        List<ScannedModule> modulesToScan = new ArrayList<>();
+        findAllClassPathJars().forEach(modulesToScan::add);
+
+        ModuleFinder moduleFinder = ModuleFinder.of(modulePaths.toArray(Path[]::new));
+        List<String> rootModules = cmdRootModules;
+        if (rootModules.contains("ALL-MODULE-PATH")) {
+            rootModules = allModuleNames(moduleFinder);
+        }
+        Configuration config = Configuration.resolveAndBind(moduleFinder, List.of(systemConfiguration()),
+                ModuleFinder.of(), rootModules);
+        for (ResolvedModule m : config.modules()) {
+            URI location = m.reference().location().orElseThrow();
+            Path path = Path.of(location);
+            checkRegularJar(path);
```

Comment:
if exploded modules aren't supported (as discussed offline), maybe the error message generated here should be more specific?

Reply:
What do you suggest? Just a note in the error message that exploded modules/class paths are not supported?

Reply:
Something like that yes

Reply:
An altermative is to use ResolvedModule::reference to get a ModuleReference, then use its open method to open the contents of the module to get a ModuleReader. That will give you a stream over the names of all resources in the module. It will work for all exploded and packaged modules.

Reply:
I took a look at this. I had to refactor a bit and introduce an intermediate abstraction (`ClassFileSource`) to abstract over classes loaded from a `JarFile` or a `ModuleReader`. At that point, it was also very easy to add a third option for directories of class files on that class path, so I've done that as well.

See: https://github.com/openjdk/jdk/pull/19774/commits/a046f6fe9465feda63609dd50014338ca218923f